Architecture.md

```diff
@@ -0,0 +1,287 @@
+We can extend the CW4 spec to enable "listeners". That is special contracts that get an update
+message each time the group changes. This works similar to the cw7 extension, but rather than make a
+special `CustomMsg` call to a native module on update, it would call a wasm contract with a
+predefined `ReceiverMsg` (the listener is sensitive and must be set by admin, as a buggy listener
+could block all group updates). Also, we only handle `sdk.AccAddress` / `HumanAddr` here, not the
+tendermint pubkey, so we are compatible with the generic cw4 groups used by multisigs.
+
+The PoE mixer is simply applies the curve `f(S, E)` to the two input groups to set an output
+group. On init, it can query both `PoS` and `PoA` and set values for every address present on either side,
+filtering out zeros, in order to get initial conditions. Afterwards, when it receives an update
+message, it can calculate the new value for that validator. On the trigger, it can query the specified
+validator address on `PoS` and `PoA` contracts, calculate `vote' = f(S', E')` and add that diff to the
+group contract it controls.
+
+A picture is worth a thousand works. Here we should how we can achieve full PoE (minus delegations)
+with only one minor additions to what we have already:
+
+```text
+   (Admin)                (Authority)                 (Distribution)
++--------------+ read   +----------------+   read   +-------------------+
++ Cw3 80% vote + -----> | Cw4/7 Group    | <------  + Reward per weight |
++--------------+        +----------------+          +-------------------+
+                             ^
+                             | write
+                       +---------------------+
+                       | PoE mixer           |
+                       +---------------------+
+                        ^                  ^
+             listener  /                    \ listener
+            +-------------+               +-------------+
+            | Cw4 Group   |               | Cw4 Group   |
+            +-------------+               +-------------+
+                 ^                               ^
+                 | r/w                           | r/w
+             (Slashing)                          |
+         +---------------------+         +---------------------+
+         | Staking Module      |         | Cw3 60% vote        |
+         |  (plus cw9 support) |         +---------------------+
+         +---------------------+
+```
+
+If we want to slash both Engagement and Stake on double-sign, we would need a second contract to call out
+to both sides and some more customization. This is doable, but one more step to add.
+
+Note that as we are adding more and more complexity to the contract composition, simple unit tests of
+contracts in isolation no longer suffice and we will have to build up much more sophisticated test harnesses
+capable of simulating the entire system in unit tests and supporting the full array of native rust tooling
```

Comment:
Maybe these are more faithfully referred to as "integration tests" at this point?

Reply:
We have been using integration tests for things like `wasmd` tests that include a compiled wasm contract inside blockchain test (or even cosmjs tests driving a blockchain node).

But yeah, I need a new term.

"Pure Rust integration tests"?

The point is if we test native rust code without compiling to wasm, we can use a whole slew of debugging and testing tools not available when we have wasm blobs in some go runtime

Reply:
> "Pure Rust integration tests"

I think this works. Or how about "pre-compilation integration tests"?